xla/service/gpu/gemm_algorithm_picker.cc

```diff
@@ -327,31 +330,45 @@ class GemmAutotuner {
         continue;
       }
 
+      num_algorithms_left_++; 
       if (!reference_algorithm) {
         TF_RETURN_IF_ERROR(stream_->Memcpy(&reference_buffer, OutputBuffer(),
                                            OutputBuffer().size()));
         reference_algorithm = profile_result.algorithm();
-      } else {
-        // Perform the comparison.
-        TF_ASSIGN_OR_RETURN(
-            bool outputs_match,
-            comparator.CompareEqual(stream_, /*current=*/OutputBuffer(),
+        continue;
+      } 
+      // Perform the comparison versus the reference algorithm.
+      TF_ASSIGN_OR_RETURN(
+          bool outputs_match,
+          comparator.CompareEqual(stream_, /*current=*/OutputBuffer(),
                                     /*expected=*/reference_buffer));
-        if (!outputs_match) {
-          LOG(ERROR) << "Results mismatch between different GEMM algorithms. "
-                     << "This is likely a bug/unexpected loss of precision.";
-          CHECK(!autotune_config_.should_crash_on_check_failure());
+      if (!outputs_match) {
+        LOG(ERROR) << "Results mismatch between different GEMM algorithms. "
+                   << "This is likely a bug/unexpected loss of precision.";
+        CHECK(!autotune_config_.should_crash_on_check_failure());
 
-          result.mutable_failure()->set_kind(AutotuneResult::WRONG_RESULT);
-          result.mutable_failure()->mutable_reference_gemm()->set_algorithm(
-              *reference_algorithm);
+        // By default, autotuner does NOT really skip wrong results, but 
+        // merely prints out the above error message: this may lead to a 
+        // great confusion. When should_skip_wrong_results() is set to true,
+        // solutions with accuracy problems will be disqualified.
+        auto kind = AutotuneResult::WRONG_RESULT;
+        if (autotune_config_.should_skip_wrong_results()) {
+          kind = AutotuneResult::DISQUALIFIED;
```

Comment:
This is not causing any tests to fail, but what if the first algorithm is incorrect? Wouldn't this disqualify all the other algorithms since they are compared with the first, when it's the first algorithm we should disqualify?

I don't know an easy way to fix this unfortunately. @tdanyluk, any ideas?

Reply:
Good question, I also asked about this in an earlier comment.
On ROCM, according to the author, the default (fallback) solution is always correct, so it sounds like this logic makes sense for their use-case.

If we really want to do this well, we could write a reference implementation that runs on CPU (such as the HLO interpreter) and compare the result with that. But the problem is that if the matrices are big, the runtime could be slow. We considered last year, that in some cases could try validating on some very sparse matrices, so that the reference computation would be fast, but it's not really easy, especially for fusions.

Reply:
Perhaps we can add an extended comment to debug_options.h where xla_gpu_autotune_level is defined ? Anyway, by default autotune_level is set to 4, hence no algorithms will be disqualified due to accuracy problems. If the user wants level 5, he should know what he is doing..

Reply:
@tdanyluk is OOO.

@pemeliya Please add a more detailed explanation of the new level 5. In particular, explain that level 5 assumes a reference algorithm is correct.

Is this actually different to the code before this PR? Looking on the left it seems that we're anyway comparing to a reference? So in that sense, the PR doesn't change this behavior? Did I miss somethig?

Reply:
@dimitar-asenov, sorry for a delay in response. Yes, I have added a more detailed description into debug_options.cc of that new level 5. As I noted in the description of that PR, currently we have the autotune  level 4 set by default. Yes, we are still comparing to a reference solution, but solutions with numerical problems **are NOT discarded** with level 4: the autotuner just prints out a warning. The rationale behind is that the reference solution might **itself be wrong**, therefore the BufferComparator test failure is merely the message for developers that there is something wrong with the underlying math library. 

Yet, in our case, on ROCM, we are sure that the first reference solution is correct (we are specifically sort the list of solutions in such a way that the default one is always on top), therefore with level 5 we would like to really discard numerically wrong solutions.